### PR TITLE
ModelWrapper

### PR DESCRIPTION
**Pull request by javdrher**
Following #60 , wrapping any kind of model defined in GPflow and influencing its input/output seems to become a shared use case. So I created a specific parent class which takes care of this, datascaler now is derived from it. This turns it into an actual implementation of the [Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern).

The class is now derived from parameterized again. Deriving from GPmodel introduced a wide range of issues which were all more or less covered but it wasn't very nice. I also noticed that for recompiling models (I tested with the modified GPflow version which is in the PR) there was some annoying split on where certain parameters were set: this is now all covered and solved issues related to using VGP.

Note that deriving from Parameterized is not as bad as it seems at first. Because all method calls not implemented in a wrapper subclass are forwarded, and an assert is placed in the constructor which verifies its actually a model we maintain the assumptions of a model through the proxy.

### CODE REVIEW

File: GPflowOpt/acquisition/acquisition.py [hunk 1]
```diff
@@ -48,7 +50,9 @@ def __init__(self, models=[], optimize_restarts=5):
         :param optimize_restarts: number of optimization restarts to use when training the models
         """
         super(Acquisition, self).__init__()
-        self._models = ParamList([DataScaler(m) for m in np.atleast_1d(models).tolist()])
+        models = np.atleast_1d(models)
+        assert all(isinstance(model, (Model, ModelWrapper))for model in models)
```
Comment (icouckuy) [inline GPflowOpt/acquisition/acquisition.py]: space before 'for' :)

File: GPflowOpt/models.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from GPflow.param import Parameterized
+from GPflow.model import Model
+
+
+class ModelWrapper(Parameterized):
+    """
+    Class for fast implementation of a wrapper for models defined in GPflow. Once wrapped, all lookups for attributes
```
Comment (icouckuy) [inline GPflowOpt/models.py]: lets make the first paragraph a one liner (so it is listed nicely if it ever becomes in the docs). So "Once..." is a new paragraph

Comment (icouckuy) [inline GPflowOpt/models.py]: Base class for fast...

File: GPflowOpt/models.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        # in the wrapped model.
+        if item.endswith('_AF_storage'):
+            method = item[1:].rstrip('_AF_storage')
+            if method in dir(self):
+                raise AttributeError("{0} has no attribute {1}".format(self.__class__.__name__, item))
+        return getattr(self.wrapped, item)
+
+    def __setattr__(self, key, value):
+        """
+        1) If setting :attr:`wrapped` attribute, point parent to this object (the datascaler).
```
Comment (icouckuy) [inline GPflowOpt/models.py]: rename the datascaler?

File: GPflowOpt/models.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                setattr(self.wrapped, key, value)
+            else:
+                #  If not, set in wrapper nonetheless.
+                super(ModelWrapper, self).__setattr__(key, value)
+
+    def __eq__(self, other):
+        return self.wrapped == other
+
+    def __str__(self, prepend=''):
+        return self.wrapped.__str__(prepend)
```
Comment (icouckuy) [inline GPflowOpt/models.py]: Useful to prepend with 'Wrapper' or something by default?

File: GPflowOpt/scaling.py [hunk 5]
```diff
@@ -12,17 +12,18 @@
...
 from .domain import UnitCube
+from .models import ModelWrapper
 
 float_type = settings.dtypes.float_type
 
 
-class DataScaler(GPModel):
+class DataScaler(ModelWrapper):
     """
     Model-wrapping class, primarily intended to assure the data in GPflow models is scaled. One DataScaler wraps one
```
Comment (icouckuy) [inline GPflowOpt/scaling.py]: oneliner?

File: GPflowOpt/models.py (hunk 4 above)
Comment (javdrher) [inline GPflowOpt/models.py]: Doesn't seem necessary, the actual method that is executed is in Parameterized so doesn't mention the model.
In the parameter naming you should see a .wrapped directive though.

Comment (javdrher) [inline GPflowOpt/models.py]: Actually, that isn't the case yet. fixing.

File: GPflowOpt/models.py [hunk 6]
```diff
@@ -49,8 +49,12 @@ def __getattr__(self, item):
 
     def __setattr__(self, key, value):
         """
-        1) If setting :attr:`wrapped` attribute, point parent to this object (the datascaler).
-        2) If setting the recompilation attribute, always do this on the wrapped class.
+        1) If setting :attr:`wrapped` attribute, point parent to this object (the ModelWrapper).
+        2) Setting attributes in the right objects. The following rules are processed in order:
+           (a) If attribute exists in wrapper, set in wrapper.
+           (b) If not object wrapped yet, set in wrapper.
```
Comment (icouckuy) [inline GPflowOpt/models.py]: no

Comment (icouckuy) [inline GPflowOpt/models.py]: what's meant with this?
that it will error out if self.wrapper is None

**Pull request merged**